### PR TITLE
feat(fogbender-element): improved simplewidget , fogbender provider and render functions

### PR DESCRIPTION
**Pull request by karannakra**
1: Improved the fogbender simple widget and  removed redundant code.
2: Implemented context and provider of component-register package and removed solid-js context and provider logic.
3: Util files changes ,  fixed render I frame and cleanup stuff.
4: Version update

### CODE REVIEW

File: packages/fogbender-element/src/utils.ts [hunk 1]
```diff
@@ -13,23 +15,35 @@ export const noopCleanup = () => {
...
+
+  if (fogbender && divRef) {
+    return renderComponent(() => fogbender.renderIframe({ headless, rootEl: divRef }));
+  } else {
+    return noopCleanup();
+  }
 };
 
 export const renderComponent = (componentRenderer: () => Promise<() => void>) => {
+  const [cleanup, setCleanup] = createSignal(() => {});
```
Comment (JLarky) [inline packages/fogbender-element/src/utils.ts:37]: does it have to be a signal?

File: packages/fogbender-element/src/FogbenderProvider.tsx [hunk 2]
```diff
@@ -1,25 +1,32 @@
...
+  "fogbender-provider",
+  { fogbender: undefined, children: undefined },
+  props => {
+    const defaultFogbender = props.fogbender || createNewFogbender();
 
-  if (!fogbender) {
-    throw new Error("No fogbender set, use FogbenderProvider to set one");
+    provide(fogbenderContext, defaultFogbender);
+
+    return [...(props.children as JSX.Element[])];
```
Comment (JLarky) [inline packages/fogbender-element/src/FogbenderProvider.tsx:28]: do we need children here?

File: packages/fogbender-element/src/FogbenderProvider.tsx [hunk 3]
```diff
@@ -1,25 +1,32 @@
...
 
 export interface FogbenderProviderProps {
-  fogbender?: Accessor<Fogbender>;
-  children?: JSX.Element;
+  fogbender?: Fogbender;
+  children?: JSX.Element[];
 }
 
-export const FogbenderContext = createContext<Accessor<Fogbender>>();
+interface Context {
```
Comment (JLarky) [inline packages/fogbender-element/src/FogbenderProvider.tsx:11]: why do we need this type?

File: packages/fogbender-element/src/utils.ts [hunk 4]
```diff
@@ -13,23 +15,35 @@ export const noopCleanup = () => {
...
 };
 
-export const renderIframe = (divRef: HTMLDivElement | undefined, headless: boolean) => {
-  const fogbender = useContext(FogbenderContext);
+export const renderIframe = (
+  divRef: HTMLDivElement | undefined,
+  headless: boolean,
+  element: ICustomElement
+) => {
+  const fogbender: Fogbender = consume(fogbenderContext, element as HTMLElement & ICustomElement);
```
Comment (JLarky) [inline packages/fogbender-element/src/utils.ts:23]: why do we need `element` here?

**Review by JLarky — CHANGES_REQUESTED** [review]

File: packages/fogbender-element/src/FogbenderProvider.tsx (hunk 3 above)
Comment (karannakra) [reply packages/fogbender-element/src/FogbenderProvider.tsx:11]: Because we are exporting context and if we were doing this without defining the interface in this file we were getting this error 
image
When in tsconfig` declaration : true`
And this was true in other packages of fogbender tsconfig files

File: packages/fogbender-element/src/utils.ts (hunk 4 above)
Comment (karannakra) [reply packages/fogbender-element/src/utils.ts:23]: Okay! So, this particular part is tricky one
image

Connected callback ran when a component is mounted on the `DOM` and  initially current element is null.
image

In connectedCallback in finally block we are assigning current element to outer element at the end which basically is undefined so when the function is executed completely current element will be undefined.

This getCurrentElement function is giving the instance of current element
image


So when we are consuming context using consume function, at the time we reach consume in simple widget current element becomes undefined as connected callback is executed completely
image

And consume is using this lookup context 
image

So long story short is lookup context needs a element ref in order to check its parent node for context.
So that's why I have pass the element from outside

File: packages/fogbender-element/src/FogbenderProvider.tsx (hunk 2 above)
Comment (karannakra) [reply packages/fogbender-element/src/FogbenderProvider.tsx:28]: Due to shadow DOM this was not able to render without props.children. Fortunately, I am now able to find a method to not opt for shadow DOM in component-register.
This children code is removed now!

File: packages/fogbender-element/src/utils.ts (hunk 4 above)
Comment (karannakra) [reply packages/fogbender-element/src/utils.ts:23]: I think was also the side effect of shadowDOM. This is also removed now.

File: packages/fogbender-element/src/utils.ts (hunk 1 above)
Comment (karannakra) [reply packages/fogbender-element/src/utils.ts:37]: I just want this to be in controlled way that's why I created this as a signal.
Normal function declaration will also do the work.

File: packages/fogbender-element/src/utils.ts (hunk 4 above)
Comment (karannakra) [reply packages/fogbender-element/src/utils.ts:23]: image

If no element is passed then when we change the props on the run time from the DOM then this error come.

**Comment by JLarky** [comment]
None of the commits except for the first one work locally for me...
